README.md
This is the standalone repo for the ERC 777 implementation of the KrypoX token

### RootRegistry

`0x20202268cb573c5c2d19c6a28185e2ec629cb1b5`

### KPX

`0x475c0ddfb60af8eac0b4fcaa9b9bdc6442362294`

### ContributionLedger

`0x58ec6d4d2b423fa90f57283fedc9537894729ec9`

### History:

#### 8/4/18

```
  Deploying RootRegistryV0_1_0...
  ... 0xb9b616303ee2557172feb0658dc64a3ecb4da0ed4fd412415a21123547c93ddb

  RootRegistryV0_1_0: 0x20202268cb573c5c2d19c6a28185e2ec629cb1b5
  ... 0x8f2f5199ca3ec032a729b10bb851e40378238beb0dded6496160372da204e7a5
  ... 0xa28514d4f5732b29b1ee4f77c2754faabc4c2044524b8d275868ae983344f756
  ... 0x4a715825e264fd6e2ac50572605bf3536331450d6b8eff983a4ab9ea91683b82
  ... 0x2df87d34a4e56ef4a1402310bcd6fe5356a0339e20693c500b038377704c58bc
```

#### 8/5/18

```
Deploying ContributionLedger...
  ... 0x432e3e926306a0c5edea8b4b4290df0ee9f9bea0b780d3717a7f2f37ddce8044
  ContributionLedger: 0x58ec6d4d2b423fa90f57283fedc9537894729ec9
```


.mainnet/history
web3.eth.getLatestBlock()
web3.eth.getBlock
web3.eth.getBlock()
web3.eth.getLatestBlock
web3.eth.getLastBlock
web3.eth.blockNumber
web3.eth.isSyncing()
web3.eth.syncing()
web3.eth.syncing
version

scripts/test.sh
#!/usr/bin/env bash

# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

# Exit script as soon as a command fails.
set -o errexit

# Executes cleanup function at script exit.
trap cleanup EXIT

cleanup() {
  # Kill the ganache instance that we started (if we started one and if it's still running).
  if [ -n "$ganache_pid" ] && ps -p $ganache_pid > /dev/null; then
    echo "Stoppping ganache instance (PID $ganache_pid)"
    kill -9 $ganache_pid
  fi
}


if [ "$SOLIDITY_COVERAGE" = true ]; then
  ganache_cmd="testrpc-sc"
  ganache_port=8555
else
  ganache_cmd="ganache-cli"
  ganache_port=8545
fi

ganache_running() {
  nc -z localhost "$ganache_port"
}

start_ganache() {
  local accounts=(                                                                                           # Account
    --account="0x0000000000000000000000000000000000000000000000000000000000000021,1000000000000000000000000" # 0x0..
    --account="0x000000000000000000000000000000000000000000000000000000000000001b,1000000000000000000000000" # 0x1..
    --account="0x00000000000000000000000000000000000000000000000000000000000000bf,1000000000000000000000000" # 0x2..
    --account="0x000000000000000000000000000000000000000000000000000000000000009e,1000000000000000000000000" # 0x3..
    --account="0x0000000000000000000000000000000000000000000000000000000000000013,1000000000000000000000000" # 0x4..
    --account="0x0000000000000000000000000000000000000000000000000000000000000058,1000000000000000000000000" # 0x5..
    --account="0x000000000000000000000000000000000000000000000000000000000000001f,1000000000000000000000000" # 0x6..
    --account="0x0000000000000000000000000000000000000000000000000000000000000023,1000000000000000000000000" # 0x7..
    --account="0x0000000000000000000000000000000000000000000000000000000000000070,1000000000000000000000000" # 0x8..
    --account="0x0000000000000000000000000000000000000000000000000000000000000019,1000000000000000000000000" # 0x9..
)
  npx ${ganache_cmd} --gasLimit 0xfffffffffff --port "$ganache_port" "${accounts[@]}" > /dev/null &
  ganache_pid=$!
  echo " (PID $ganache_pid)"
}

if ganache_running; then
  echo "Using existing ganache instance"
else
  printf "Starting new ganache instance"
  start_ganache
fi

if [ "$SOLIDITY_COVERAGE" = true ]; then
  npx solidity-coverage
else
  npx truffle test "$@"
fi

test/KPX.test.js
/* This Source Code Form is subject to the terms of the Mozilla Public
* License, v. 2.0. If a copy of the MPL was not distributed with this
* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
const abi = require('ethereumjs-abi');
const chai = require('chai');
const assert = chai.assert;
chai.use(require('chai-as-promised')).should();
const { URL } = require('url');
const Web3 = require('web3');
const OldReferenceToken = artifacts.require('KPXV0_1_0');
const OldEIP820Registry = artifacts.require('RootRegistryV0_1_0');
const OldUnstructuredOwnedUpgradeabilityProxy = artifacts.require(
  'UnstructuredOwnedUpgradeabilityProxy'
);
const utils = require('./utils');

function encodeCall(name, _arguments, values) {
  const methodId = abi.methodID(name, _arguments).toString('hex');
  const params = abi.rawEncode(_arguments, values).toString('hex');
  return `0x${methodId}${params}`;
}

contract('KPXV0_1_0', function(accounts) {
  const provider = new URL(this.web3.currentProvider.host);
  provider.protocol = 'ws';
  const web3 = new Web3(provider.toString());
  const ReferenceToken = new web3.eth.Contract(OldReferenceToken.abi, {
    data: OldReferenceToken.bytecode,
  });
  let UnstructuredOwnedUpgradeabilityProxy;

  let Erc820Registry = new web3.eth.Contract(OldEIP820Registry.abi, {
    data: OldEIP820Registry.bytecode,
    arguments: [accounts[0]],
  });

  let token = {
    name: 'KryptoX',
    symbol: 'KPX',
    granularity: '0.01',
    totalSupply: '0',
    defaultBalance: '0',
  };

  const deployContract = ReferenceToken.deploy();

  const initializeToken = async (token, registryAddr) => {
    await token.contract.methods
      .initialize(
        token.name,
        token.symbol,
        web3.utils.toWei(token.granularity),
        registryAddr
      )
      .send({
        from: accounts[0],
        gas: 4600000,
      });
  };

  after(async function() {
    await web3.currentProvider.connection.close();
  });

  beforeEach(async function() {
    Erc820Registry = await Erc820Registry.deploy({
      arguments: [accounts[0]],
    }).send({
      from: accounts[0],
      gas: 4600000,
    });

    UnstructuredOwnedUpgradeabilityProxy = new web3.eth.Contract(
      OldUnstructuredOwnedUpgradeabilityProxy.abi,
      {
        data: OldUnstructuredOwnedUpgradeabilityProxy.bytecode,
        arguments: [Erc820Registry.options.address],
      }
    );

    UnstructuredOwnedUpgradeabilityProxy = await UnstructuredOwnedUpgradeabilityProxy.deploy(
      {
        arguments: [Erc820Registry.options.address],
      }
    ).send({
      from: accounts[0],
      gas: 4600000,
    });
    await Erc820Registry.methods
      .setProxy(
        UnstructuredOwnedUpgradeabilityProxy.options.address,
        'KPX',
        true
      )
      .send({
        from: accounts[0],
        gas: 4600000,
      });

    // // Use Web3.js 1.0
    const estimateGas = await deployContract.estimateGas();
    token.contract = await deployContract.send({
      from: accounts[0],
      gasLimit: estimateGas,
    });

    const tokenInitParamTypes = [
      'string',
      'string',
      'uint256',
      'address',
      'address',
    ];
    const tokenInitParamVals = [
      'KryptoX',
      'KPX',
      web3.utils.toWei('0.01'),
      Erc820Registry.options.address,
      accounts[0],
    ];
    const initializeData = encodeCall(
      'initialize',
      tokenInitParamTypes,
      tokenInitParamVals
    );
    await UnstructuredOwnedUpgradeabilityProxy.methods
      .upgradeToAndCall(
        'KPX',
        '0_1_0',
        token.contract.options.address,
        initializeData
      )
      .send({
        from: accounts[0],
        gasLimit: estimateGas,
      });

    token.contract = new web3.eth.Contract(
      OldReferenceToken.abi,
      UnstructuredOwnedUpgradeabilityProxy.options.address,
      {
        data: OldReferenceToken.bytecode,
      }
    );

    assert.ok(token.contract.options.address);

    token.disableERC20 = async function() {
      await token.contract.methods
        .disableERC20()
        .send({ gas: 300000, from: accounts[0] });
    };

    token.genMintTxForAccount = function(account, amount, operator, gas) {
      return token.contract.methods
        .mint(account, web3.utils.toWei(amount), '0xcafe')
        .send.request({ gas: gas, from: operator });
    };
  });

  describe('Creation', function() {
    it('should not be able to initalize the token again', async function() {
      await token.contract.methods
        .initialize(
          token.name,
          token.symbol,
          web3.utils.toWei(token.granularity),
          Erc820Registry.options.address,
          accounts[0]
        )
        .send({ gas: 300000, from: accounts[0] })
        .should.be.rejectedWith('revert');
    });
  });

  require('./utils/attributes').test(web3, accounts, token);
  require('./utils/mint').test(web3, accounts, token);
  require('./utils/burn').test(web3, accounts, token);
  require('./utils/send').test(web3, accounts, token);
  require('./utils/operator').test(web3, accounts, token);
  require('./utils/tokensSender').test(web3, accounts, token);
  require('./utils/tokensRecipient').test(web3, accounts, token);
  require('./utils/erc20Compatibility').test(web3, accounts, token);
  require('./utils/pause').test(web3, accounts, token);
  require('./utils/lockAndDistribute').test(web3, accounts, token);

  describe('ERC20 Disable', function() {
    it('should disable ERC20 compatibility', async function() {
      let erc820Registry = utils.getERC820Registry(
        web3,
        Erc820Registry.options.address
      );
      let erc20Hash = web3.utils.keccak256('ERC20Token');
      let erc20Addr = await erc820Registry.methods
        .getInterfaceImplementer(token.contract.options.address, erc20Hash)
        .call();

      assert.strictEqual(erc20Addr, token.contract.options.address);

      await token.disableERC20();
      console.log(accounts[0], await token.contract.methods.owner().call());
      await utils.getBlock(web3);
      erc20Addr = await erc820Registry.methods
        .getInterfaceImplementer(token.contract.options.address, erc20Hash)
        .call();

      assert.strictEqual(
        erc20Addr,
        '0x0000000000000000000000000000000000000000'
      );
    });
  });

  require('./utils/erc20Disabled').test(web3, accounts, token);
});


test/utils/mint.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
const chai = require('chai');
chai.use(require('chai-as-promised')).should();
const utils = require('./index');

exports.test = function(web3, accounts, token) {
  describe('minting', function() {
    it(`should mint 10 ${token.symbol} for ${utils.formatAccount(
      accounts[1]
    )}`, async function() {
      await utils.assertBalance(web3, token, accounts[1], 0);

      await token.contract.methods
        .mint(accounts[1], web3.utils.toWei('10'), '0x')
        .send({ gas: 300000, from: accounts[0] });

      await utils.getBlock(web3);

      await utils.assertTotalSupply(web3, token, 10);
      await utils.assertBalance(web3, token, accounts[1], 10);
    });

    it(
      `should mint 10 ${token.symbol} for ` +
        `${utils.formatAccount(accounts[1])} (ERC20 disabled)`,
      async function() {
        await utils.assertBalance(web3, token, accounts[1], 0);

        await token.disableERC20();

        await token.contract.methods
          .mint(accounts[1], web3.utils.toWei('10'), '0x')
          .send({ gas: 300000, from: accounts[0] });

        await utils.getBlock(web3);

        // TODO check events
        await utils.assertTotalSupply(web3, token, 10);
        await utils.assertBalance(web3, token, accounts[1], 10);
      }
    );

    it(`should not mint -10 ${
      token.symbol
    } (negative amount)`, async function() {
      await utils.assertBalance(web3, token, accounts[1], 0);

      await token.contract.methods
        .mint(accounts[1], web3.utils.toWei('-10'), '0x')
        .send({ gas: 300000, from: accounts[0] })
        .should.be.rejectedWith('revert');

      await utils.getBlock(web3);

      await utils.assertTotalSupply(web3, token, 0);
      await utils.assertBalance(web3, token, accounts[1], 0);
    });

    it(`should not mint 0.007 ${
      token.symbol
    } (< granulairty)`, async function() {
      await utils.assertBalance(web3, token, accounts[1], 0);

      await token.contract.methods
        .mint(accounts[1], web3.utils.toWei('0.007'), '0x')
        .send({ gas: 300000, from: accounts[0] })
        .should.be.rejectedWith('revert');

      await utils.getBlock(web3);
      await utils.assertBalance(web3, token, accounts[1], 0);
      await utils.assertTotalSupply(web3, token, 0);
    });
  });
};


migrations/1_deploy_KPX.js
const { upgradeAndMigrateContracts } = require('../test/helpers/contracts');
const { kpxConfig } = require('../test/helpers/contractConfigs');
const RootRegistryV0_1_0 = artifacts.require('RootRegistryV0_1_0');
const Web3 = require('web3');

module.exports = (deployer, network, accounts) => {
  deployer.then(async () => {
    // web3 = new Web3(web3.currentProvider);
    // const config = {
    //   network,
    //   artifacts,
    //   web3,
    //   accounts,
    //   deployer,
    // };
    // const root = await deployer.deploy(
    //   RootRegistryV0_1_0,
    //   '0x1c128de6b93557651e0dc5b55f23496ffac7a0a9',
    //   { gasPrice: 20000000000 }
    // );
    // const contractsToUpgrade = [kpxConfig];
    // const deployedContracts = await upgradeAndMigrateContracts(
    //   config,
    //   '0x1c128de6b93557651e0dc5b55f23496ffac7a0a9',
    //   contractsToUpgrade,
    //   '0x1c128de6b93557651e0dc5b55f23496ffac7a0a9',
    //   root
    // );
    // const kpx = deployedContracts[0];
    // console.log(
    //   'kpx.upgradeableContractAtProxy.address:',
    //   kpx.upgradeableContractAtProxy.address
    // );
  });
};

// module.exports = (deployer, network) => {
//   deployer.then(async () => {
//     global.network = network;
//     // yeah we won't be using this...
//   });
// };

//const { upgradeAndMigrateContracts } = require('../test/helpers/contracts');
// const { kpxConfig } = require('../test/helpers/contractConfigs');
// const RootRegistryV0_1_0 = artifacts.require('RootRegistryV0_1_0');
// const Web3 = require('web3');

// module.exports = (deployer, network, accounts) => {
//   deployer.then(async () => {
//     web3 = new Web3(web3.currentProvider);
//     console.log(accounts[0]);

//     const config = {
//       network,
//       artifacts,
//       web3,
//       accounts,
//       deployer,
//     };
//     const root = await deployer.deploy(RootRegistryV0_1_0, accounts[0]);

//     const contractsToUpgrade = [kpxConfig];
//     const deployedContracts = await upgradeAndMigrateContracts(
//       config,
//       '0x1c128de6b93557651e0dc5b55f23496ffac7a0a9',
//       contractsToUpgrade,
//       '0x1c128de6b93557651e0dc5b55f23496ffac7a0a9',
//       root
//     );
//     const kpx = deployedContracts[0];
//     console.log(
//       'kpx.upgradeableContractAtProxy.address',
//       kpx.upgradeableContractAtProxy.address
//     );
//   });
// };

// // module.exports = (deployer, network) => {
// //   deployer.then(async () => {
// //     global.network = network;
// //     // yeah we won't be using this...
// //   });
// // };


truffle.js
// require('babel-register')({
//   ignore: /node_modules\/(?!zeppelin-solidity)/,
// });
// require('babel-polyfill');

/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
module.exports = {
  networks: {
    development: {
      host: 'localhost',
      port: 8545,
      network_id: '*', // eslint-disable-line camelcase
    },
    test: {
      host: 'localhost',
      port: 8545,
      network_id: '*', // eslint-disable-line camelcase
    },
    coverage: {
      host: 'localhost',
      port: 8555,
      network_id: '*', // eslint-disable-line camelcase
      gas: 0xfffffffffff,
      gasPrice: 0x01,
    },
    ganache: {
      host: 'localhost',
      port: 8545,
      network_id: '*', // eslint-disable-line camelcase
    },
    ropstenGeth: {
      host: 'localhost',
      port: 8545,
      network_id: 3,
      gas: 5219725,
      from: '0x1c128de6b93557651e0dc5b55f23496ffac7a0a9',
    },
    live: {
      gasPrice: 50000000000, // 50 gwei
      host: 'localhost',
      port: 8545,
      network_id: 1,
      gas: 5219725,
      from: '0x1c128de6b93557651e0dc5b55f23496ffac7a0a9',
    },
  },
};
